pcr_profile.go

```diff
@@ -0,0 +1,162 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package secboot
+
+import (
+	"fmt"
+
+	"github.com/chrisccoulson/go-tpm2"
+
+	"golang.org/x/xerrors"
+)
+
+// pcrValuesList is a list of PCR value combinations computed from PCRProtectionProfile.
+type pcrValuesList []tpm2.PCRValues
+
+// setValue sets the specified PCR to the supplied value for all branches.
+func (l pcrValuesList) setValue(alg tpm2.HashAlgorithmId, pcr int, value tpm2.Digest) {
+	for _, v := range l {
+		v.SetValue(pcr, alg, value)
+	}
+}
+
+// extendValue extends the specified PCR with the supplied value for all branches.
+func (l pcrValuesList) extendValue(alg tpm2.HashAlgorithmId, pcr int, value tpm2.Digest) {
+	for _, v := range l {
+		if _, ok := v[alg]; !ok {
+			v[alg] = make(map[int]tpm2.Digest)
+		}
+		if _, ok := v[alg][pcr]; !ok {
+			v[alg][pcr] = make(tpm2.Digest, alg.Size())
+		}
+		h := alg.NewHash()
+		h.Write(v[alg][pcr])
+		h.Write(value)
+		v[alg][pcr] = h.Sum(nil)
+	}
+}
+
+func (l pcrValuesList) copy() (out pcrValuesList) {
+	for _, v := range l {
+		ov := make(tpm2.PCRValues)
+		for alg := range v {
+			ov[alg] = make(map[int]tpm2.Digest)
+			for pcr := range v[alg] {
+				ov[alg][pcr] = v[alg][pcr]
+			}
+		}
+		out = append(out, ov)
+	}
+	return
+}
+
+type pcrProfileOp int
+
+const (
+	pcrProfileOpAdd pcrProfileOp = iota
+	pcrProfileOpAddFromTPM
+	pcrProfileOpExtend
+	pcrProfileOpOr
+)
+
+// pcrProtectionProfileInstr is a building block of PCRProtectionProfile.
+type pcrProtectionProfileInstr struct {
+	op   pcrProfileOp
+	args []interface{}
+}
+
+// PCRProtectionProfile defines the PCR profile used to protect a key sealed with SealKeyToTPM. It contains a sequence of instructions
+// for computing combinations of PCR values that a key will be protected against. The profile is built using the methods of this type.
+type PCRProtectionProfile []pcrProtectionProfileInstr
```

Comment:
I find the naked unexported slice confusing, given how slice works (they are not lisp  cons) you cannot really build trees with them so the sliceness is not very useful, I think this should be wrapped in a struct, and the method below become mutating-in-place